text/0517-io-os-reform.md

```diff
@@ -473,12 +650,346 @@ throughout IO, we can go on to explore the modules in detail.
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+impl Read for Repeat { type Err = Void; ... }
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+impl Write for Sink { type Err = Void; ... }
+
+// Copies all data from a `Read` to a `Write`, returning the amount of data
+// copied.
+pub fn copy<E, R, W>(r: &mut R, w: &mut W) -> Result<u64, E>
+    where R: Read<Err = E>,
+          W: Write<Err = E>,
+          R::Err: WriteAllError
+```
+
+Like `write_all`, the `copy` method will discard the amount of data already
+written on any error and also discard any partially read data on a `write`
+error. This method is intended to be a convenience and `write` should be used
+directly if this is not desirable.
+
+#### Void
+[Void]: #void
+
+A new concrete error type will be added in the standard library. A new module
+`std::void` will be introduced with the following contents:
+
+```rust
+pub enum Void {}
+
+impl<E: Error> FromError<Void> for E {
+    fn from_error(v: Void) -> E {
+        match v {}
+    }
+}
+```
+
+Applications for an uninhabited enum have come up from time-to-time, and some of
+the core I/O adapters represent a fairly concrete use case motivating its
+existence. By using an associated `Err` type of `Void`, each I/O object is
+indicating that it *can never fail*. This allows the types themselves to be more
+optimized in the future as well as enabling interoperation with many other error
+types via the `map_err` adaptor.
+
+Some possible future optimizations include:
+
+* `Result<T, Void>` could be represented in memory exactly as `T` (no
+  discriminant).
+* The `unused_must_use` lint could understand that `Result<T, Void>` does not
+  need to be warned about.
+
+This RFC does not propose implementing these modifications at this time,
+however.
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is removed and the `seek` signature is refactored with more precise
+types:
+
+```rust
+pub trait Seek {
+    type Err;
+    // returns the new position after seeking
+    fn seek(&mut self, pos: SeekPos) -> Result<u64, Err>;
+}
+
+pub enum SeekPos {
+    FromStart(u64),
+    FromEnd(i64),
+    FromCur(i64),
+}
+```
+
+The old `tell` function can be regained via `seek(SeekPos::FromCur(0))`.
+
+#### Buffering
+[Buffering]: #buffering
+
+The current `Buffer` trait will be renamed to `BufferedRead` for
+clarity (and to open the door to `BufferedWrite` at some later
+point):
+
+```rust
+pub trait BufferedRead: Read {
+    fn fill_buf(&mut self) -> Result<&[u8], Self::Err>;
+    fn consume(&mut self, amt: uint);
+}
+
+pub trait BufferedReadExt: BufferedRead {
+    fn read_until(&mut self, byte: u8) -> ReadUntil { ... }
+    fn lines(&mut self) -> Lines<Self, Self::Err> { ... };
+}
+```
+
+where `ReadUntil: Iterator<Result<Vec<u8>, Self::Err>>`.  In addition,
```

Comment:
As of now, read_until returns a Vec<u8> including the separator, except if EOF occurred. Will the ReadUntil iterator change this behavior? (to be like the Lines iterator from std::str instead of the Lines iterator from std::io)